### DISCUSSION TITLE
SDIO support for STM32

### DISCUSSION THREAD

**Discussion by GentlemanXu** [modm-io/modm#778] [Ideas]
I have been using modm in our robotics project for a while and I'm trying to figure out how to read and write logs to SDCard with FATFS. I'm currently using an STM32F4 which supports 4-bits mode SDIO and DMA access, but I couldn't find anything related to SDCard, either SDIO support or SDCard support through SPI. I would like to implement support for SDCard under both SPI and SDIO, polling or with DMA, but I found integrating 3000 lines ST HAL SDIO module into modm challenging. I also found an [alternative SDIO implementation from RT-Thread](https://github.com/RT-Thread-packages/stm32_sdio). Any suggestions?

**Comment by salkinium** [comment]
We haven't gotten there yet, sorry. Haven't looked into what's necessary to integrate ST HAL into modm either.

I can point to the resources on http://elm-chan.org/fsw/ff/00index_e.html.

**Reply by GentlemanXu** [reply]
What's the general methodology or policy for integrating third party lib like ST HAL? Maybe an option or module, something like modm::st_f4_hal?

**Reply by salkinium** [reply]
It would be an external repository, since the ST HAL is way too large to include into modm. I am interested in at least providing a way to use the ST HAL for peripherals we don't yet support.

I assume that there are probably only a few things to port (like the 1ms tick the HAL uses) and they will probably be the same across all families (I assume there is a lot of duplicated code, since the families share a lot of identical/similiar peripherals).

I would start with just including the ST HAL code into your modm project to find out what breaks and then try to figure out what needs to be ported and what changes are needed in modm to support that.
